### PR TITLE
api: adapt rolling stock model adding profiles

### LINKED ISSUE #2240: Adapt rolling model for electric profiles

### LINKED ISSUE #2264: Implement an effort curve by comfort mode during simulations
## User

Study officer 

## Description

Use during the running time calculation the effort curve modified in fonction of the chosen comfort mode  and electrical profile.

## Goal

To improve the simulations

## Acceptance criteria

- [ ] Once that the infrastructure and the rolling stock with its comfort have been selected, the right effort curve is used 
- [ ] When the train runs on an infrastructure zone electrified by 1500 V or by 25 kV or not electrified, the set of curves corresponding to this traction mode is used
- [ ] The curve within the set corresponding to the comfort mode chosen is used
- [ ] If we don't dispose of the curve corresponding to the selected mode, the 'nominal curve' (SC) curve is chosen 
- [ ] If no comfort condition is selected, the SC curve should be chosen
- [ ] if there is no SC curve, the ? curve should be chosen

## Definition of ready

- [ ] The user story is well defined
- [ ] Acceptance criterias are defined
- [ ] The individual responsible for validating the user story is assigned
- [ ] The team should understand how to demonstrate the features
- [ ] Does not depend on an missing feature
- [ ] Negociable with the implementation team
- [ ] Valuable to the end users
- [ ] Estimable by the implementation team
- [ ] As small as it can be
- [ ] Testable by the implementation team

## Definition of done

- [ ] Acceptance criteria are met
- [ ] Automated tests are written against acceptance criteria
- [ ] Project deployed to the testing environment
- [ ] Story ok-ed by Product Owner
- [ ] Documentation updated

### PR DESCRIPTION
**Pull request by flomonster**
- New rolling stock model v3.0
- Adapt railjson to 3.0.1
  - Catenaries voltages moved from `float` to `str`
- Handle multiple effort curves
- Handle train comfort

### CODE REVIEW

File: api/openapi.yaml [hunk 1]
```diff
@@ -1172,6 +1172,11 @@ components:
               speed_limit_composition:
                 description: Train category for speed limits
                 type: string
+              comfort:
+                description: Train category for speed limits
```
Comment (younesschrifi) [inline api/openapi.yaml]: The description seems to be false. It should maybe be : Comfort category

File: api/osrd_infra/migrations/0009_rolling_stocks_v3_0.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+# Generated by Django 4.0.8 on 2022-11-04 10:24
+
+from django.db import migrations, models
+
+import osrd_infra.schemas.rolling_stock
+import osrd_infra.utils
+
+
+def apply_migration(apps, schema_editor):
```
Comment (younesschrifi) [inline api/osrd_infra/migrations/0009_rolling_stocks_v3_0.py:9]: arg schema_editor is unused

File: api/osrd_infra/migrations/0009_rolling_stocks_v3_0.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            infra.railjson_version = "3.0.1"
+            updated_infra.append(infra)
+            for catenary in CatenaryModel.objects.all():
+                catenary.data["voltage"] = str(int(catenary.data["voltage"]))
+                updated_catenaries.append(catenary)
+    Infra.objects.bulk_update(updated_infra, ["railjson_version"])
+    CatenaryModel.objects.bulk_update(updated_catenaries, ["data"])
+
+
+def revert_migration(apps, schema_editor):
```
Comment (younesschrifi) [inline api/osrd_infra/migrations/0009_rolling_stocks_v3_0.py:27]: arg schema_editor is unused

File: api/osrd_infra/migrations/0009_rolling_stocks_v3_0.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                if "15" in catenary.data["voltage"]:
+                    catenary.data["voltage"] = 1500.0
+                else:
+                    catenary.data["voltage"] = 25_000.0
+                updated_catenaries.append(catenary)
+    Infra.objects.bulk_update(updated_infra, ["railjson_version"])
+    CatenaryModel.objects.bulk_update(updated_catenaries, ["data"])
+
+
+def clear_rolling_stocks(apps, schema_editor):
```
Comment (younesschrifi) [inline api/osrd_infra/migrations/0009_rolling_stocks_v3_0.py:48]: arg schema_editor is unused

File: api/static/example_rolling_stock.json [hunk 5]
```diff
@@ -1,72 +1 @@
...
-    "rolling_resistance": {
-        "type": "davis",
-        "A": 5400.0,
-        "B": 200.0,
-        "C": 12.0
-    },
-    "electric_only": false,
-    "compatible_voltages": []
-}
+{"version": "3.0", "length": 400, "max_speed": 80, "startup_time": 10, "startup_acceleration": 0.05, "comfort_acceleration": 0.25, "inertia_coefficient": 1.05, "features": ["TVM300", "TVM430", "ETCS1", "ETCS2", "KVB"], "name": "fast_rolling_stock", "mass": 900000, "loading_gauge": "G1", "gamma": {"value": 0.5, "type": "CONST"}, "rolling_resistance": {"type": "davis", "A": 5400.0, "B": 200.0, "C": 12.0}, "effort_curves": {"default_mode": "thermal", "modes": {"thermal": {"curves": [], "default_curve": {"speeds": [0.0, 5.0, 10.0, 15.0, 20.0, 22.0, 27.0, 32.0, 37.0, 42.0, 47.0, 52.0, 57.0, 62.0, 67.0, 72.0, 77.0], "max_efforts": [441666.6666666667, 439473.6842105263, 435714.28571428574, 427777.77777777775, 400000.0, 350971.5388299929, 347206.93642395496, 346938.7385068534, 344395.0325320009, 334314.2138640166, 313589.8108101956, 283584.5657113532, 250604.14937613969, 222698.71360301683, 204685.35097358702, 195984.55717992093, 192916.7642524637]}, "is_electric": false}}}}
```
Comment (younesschrifi) [inline api/static/example_rolling_stock.json]: You should use a formatter here

File: core/examples/rolling_stocks/fast_rolling_stock.json [hunk 6]
```diff
@@ -1,72 +1 @@
...
-        "type": "davis",
-        "A": 5400.0,
-        "B": 200.0,
-        "C": 12.0
-    },
-    "compatible_voltages": [],
-    "electric_only": false
-}
\ No newline at end of file
+{"version": "3.0", "length": 400, "max_speed": 80, "startup_time": 10, "startup_acceleration": 0.05, "comfort_acceleration": 0.25, "inertia_coefficient": 1.05, "features": ["TVM300", "TVM430", "ETCS1", "ETCS2", "KVB"], "name": "fast_rolling_stock", "mass": 900000, "loading_gauge": "G1", "gamma": {"value": 0.5, "type": "CONST"}, "rolling_resistance": {"type": "davis", "A": 5400.0, "B": 200.0, "C": 12.0}, "effort_curves": {"default_mode": "thermal", "modes": {"thermal": {"curves": [], "default_curve": {"speeds": [0.0, 5.0, 10.0, 15.0, 20.0, 22.0, 27.0, 32.0, 37.0, 42.0, 47.0, 52.0, 57.0, 62.0, 67.0, 72.0, 77.0], "max_efforts": [441666.6666666667, 439473.6842105263, 435714.28571428574, 427777.77777777775, 400000.0, 350971.5388299929, 347206.93642395496, 346938.7385068534, 344395.0325320009, 334314.2138640166, 313589.8108101956, 283584.5657113532, 250604.14937613969, 222698.71360301683, 204685.35097358702, 195984.55717992093, 192916.7642524637]}, "is_electric": false}}}}
```
Comment (younesschrifi) [inline core/examples/rolling_stocks/fast_rolling_stock.json]: You should use a formatter

File: core/examples/rolling_stocks/fast_rolling_stock_high_gamma.json [hunk 7]
```diff
@@ -1,72 +1 @@
...
-        "type": "davis",
-        "A": 5400.0,
-        "B": 200.0,
-        "C": 12.0
-    },
-    "compatible_voltages": [],
-    "electric_only": false
-}
\ No newline at end of file
+{"version": "3.0", "length": 400, "max_speed": 80, "startup_time": 10, "startup_acceleration": 0.05, "comfort_acceleration": 0.25, "inertia_coefficient": 1.05, "features": ["TVM300", "TVM430", "ETCS1", "ETCS2", "KVB"], "name": "fast_rolling_stock_high_gamma", "mass": 900000, "loading_gauge": "G1", "gamma": {"value": 0.95, "type": "CONST"}, "rolling_resistance": {"type": "davis", "A": 5400.0, "B": 200.0, "C": 12.0}, "effort_curves": {"default_mode": "thermal", "modes": {"thermal": {"curves": [], "default_curve": {"speeds": [0.0, 5.0, 10.0, 15.0, 20.0, 22.0, 27.0, 32.0, 37.0, 42.0, 47.0, 52.0, 57.0, 62.0, 67.0, 72.0, 77.0], "max_efforts": [441666.6666666667, 439473.6842105263, 435714.28571428574, 427777.77777777775, 400000.0, 350971.5388299929, 347206.93642395496, 346938.7385068534, 344395.0325320009, 334314.2138640166, 313589.8108101956, 283584.5657113532, 250604.14937613969, 222698.71360301683, 204685.35097358702, 195984.55717992093, 192916.7642524637]}, "is_electric": false}}}}
```
Comment (younesschrifi) [inline core/examples/rolling_stocks/fast_rolling_stock_high_gamma.json]: You should use a formatter

File: core/examples/rolling_stocks/short_fast_rolling_stock.json [hunk 8]
```diff
@@ -1,72 +1 @@
...
-        "type": "davis",
-        "A": 5400.0,
-        "B": 200.0,
-        "C": 12.0
-    },
-    "compatible_voltages": [],
-    "electric_only": false
-}
\ No newline at end of file
+{"version": "3.0", "length": 15, "max_speed": 20, "startup_time": 10, "startup_acceleration": 0.05, "comfort_acceleration": 0.25, "inertia_coefficient": 1.05, "features": ["TVM300", "TVM430", "ETCS1", "ETCS2", "KVB"], "name": "short_fast_rolling_stock", "mass": 900000, "loading_gauge": "G1", "gamma": {"value": 0.5, "type": "CONST"}, "rolling_resistance": {"type": "davis", "A": 5400.0, "B": 200.0, "C": 12.0}, "effort_curves": {"default_mode": "thermal", "modes": {"thermal": {"curves": [], "default_curve": {"speeds": [0.0, 5.0, 10.0, 15.0, 20.0, 22.0, 27.0, 32.0, 37.0, 42.0, 47.0, 52.0, 57.0, 62.0, 67.0, 72.0, 77.0], "max_efforts": [441666.6666666667, 439473.6842105263, 435714.28571428574, 427777.77777777775, 400000.0, 350971.5388299929, 347206.93642395496, 346938.7385068534, 344395.0325320009, 334314.2138640166, 313589.8108101956, 283584.5657113532, 250604.14937613969, 222698.71360301683, 204685.35097358702, 195984.55717992093, 192916.7642524637]}, "is_electric": false}}}}
```
Comment (younesschrifi) [inline core/examples/rolling_stocks/short_fast_rolling_stock.json]: You should use a formatter

File: core/examples/rolling_stocks/short_slow_rolling_stock.json [hunk 9]
```diff
@@ -1,72 +1 @@
...
-        "type": "davis",
-        "A": 5400.0,
-        "B": 200.0,
-        "C": 12.0
-    },
-    "compatible_voltages": [],
-    "electric_only": false
-}
\ No newline at end of file
+{"version": "3.0", "length": 15, "max_speed": 10, "startup_time": 10, "startup_acceleration": 0.05, "comfort_acceleration": 0.25, "inertia_coefficient": 1.05, "features": ["TVM300", "TVM430", "ETCS1", "ETCS2", "KVB"], "name": "short_slow_rolling_stock", "mass": 900000, "loading_gauge": "G1", "gamma": {"value": 0.5, "type": "CONST"}, "rolling_resistance": {"type": "davis", "A": 5400.0, "B": 200.0, "C": 12.0}, "effort_curves": {"default_mode": "thermal", "modes": {"thermal": {"curves": [], "default_curve": {"speeds": [0.0, 5.0, 10.0, 15.0, 20.0, 22.0, 27.0, 32.0, 37.0, 42.0, 47.0, 52.0, 57.0, 62.0, 67.0, 72.0, 77.0], "max_efforts": [441666.6666666667, 439473.6842105263, 435714.28571428574, 427777.77777777775, 400000.0, 350971.5388299929, 347206.93642395496, 346938.7385068534, 344395.0325320009, 334314.2138640166, 313589.8108101956, 283584.5657113532, 250604.14937613969, 222698.71360301683, 204685.35097358702, 195984.55717992093, 192916.7642524637]}, "is_electric": false}}}}
```
Comment (younesschrifi) [inline core/examples/rolling_stocks/short_slow_rolling_stock.json]: Formatter here too

File: core/examples/rolling_stocks/slow_rolling_stock.json [hunk 10]
```diff
@@ -1,72 +1 @@
...
-        "type": "davis",
-        "A": 5400.0,
-        "B": 200.0,
-        "C": 12.0
-    },
-    "compatible_voltages": [],
-    "electric_only": false
-}
\ No newline at end of file
+{"version": "3.0", "length": 400, "max_speed": 10, "startup_time": 10, "startup_acceleration": 0.05, "comfort_acceleration": 0.25, "inertia_coefficient": 1.05, "features": ["TVM300", "TVM430", "ETCS1", "ETCS2", "KVB"], "name": "slow_rolling_stock", "mass": 900000, "loading_gauge": "G1", "gamma": {"value": 0.5, "type": "CONST"}, "rolling_resistance": {"type": "davis", "A": 5400.0, "B": 200.0, "C": 12.0}, "effort_curves": {"default_mode": "thermal", "modes": {"thermal": {"curves": [], "default_curve": {"speeds": [0.0, 5.0, 10.0, 15.0, 20.0, 22.0, 27.0, 32.0, 37.0, 42.0, 47.0, 52.0, 57.0, 62.0, 67.0, 72.0, 77.0], "max_efforts": [441666.6666666667, 439473.6842105263, 435714.28571428574, 427777.77777777775, 400000.0, 350971.5388299929, 347206.93642395496, 346938.7385068534, 344395.0325320009, 334314.2138640166, 313589.8108101956, 283584.5657113532, 250604.14937613969, 222698.71360301683, 204685.35097358702, 195984.55717992093, 192916.7642524637]}, "is_electric": false}}}}
```
Comment (younesschrifi) [inline core/examples/rolling_stocks/slow_rolling_stock.json]: Formatter

File: core/src/main/java/fr/sncf/osrd/railjson/parser/RJSRollingStockParser.java [hunk 11]
```diff
@@ -92,15 +106,49 @@ private static RJSRollingResistance.Davis parseRollingResistance(
...
+        return new RollingStock.EffortCurveConditions(parseComfort(rjsCond.comfort));
+    }
+
+    /** Parse rjsComfort into a RollingStock comfort */
+    public static RollingStock.Comfort parseComfort(RJSComfortType rjsComfort) {
+        if (rjsComfort == RJSComfortType.AC)
+            return RollingStock.Comfort.AC;
+        if (rjsComfort == RJSComfortType.HEATING)
+            return RollingStock.Comfort.HEATING;
+        return RollingStock.Comfort.STANDARD;
```
Comment (younesschrifi) [inline core/src/main/java/fr/sncf/osrd/railjson/parser/RJSRollingStockParser.java:127]:
```suggestion
            return RollingStock.Comfort.HEATING;
        else 
            return RollingStock.Comfort.STANDARD;
```

**Review by younesschrifi — COMMENTED** [review]
Maybe you should have less commits

File: api/osrd_infra/migrations/0009_rolling_stocks_v3_0.py (hunk 3 above)
Comment (flomonster) [reply api/osrd_infra/migrations/0009_rolling_stocks_v3_0.py:27]: I can not change the signature of this funtion. It must match the signature required by django `RunPython` migration.
https://docs.djangoproject.com/en/4.1/ref/migration-operations/#django.db.migrations.operations.RunPython

File: core/src/main/java/fr/sncf/osrd/railjson/parser/RJSRollingStockParser.java (hunk 11 above)
Comment (flomonster) [reply core/src/main/java/fr/sncf/osrd/railjson/parser/RJSRollingStockParser.java:127]: It is not necessary to put an `else` here.

File: api/osrd_infra/schemas/infra.py [hunk 12]
```diff
@@ -159,7 +159,7 @@ def length(self) -> float:
 
 class ApplicableDirectionsTrackRange(BaseModel):
     """
-    This class is used to define track ranges that are associated with certain classes in the infrastructure.
+    An applicable directions track range is a track range with an associated direction.
```
Comment (eckter) [inline api/osrd_infra/schemas/infra.py]:
```suggestion
    An applicable directions track range is a track range with associated directions.
```

File: api/osrd_infra/schemas/infra.py [hunk 13]
```diff
@@ -224,17 +225,18 @@ class SwitchPortConnection(BaseModel):
 
 class SwitchType(BaseObjectTrait):
     """
-    This class is used to define switch types.
-    We always have three switch types characterized by their identifier: point, cross and double cross.
+    Switch types are used to define what kind of switch is used in the infrastructure.
+    A switch type is define by a list of ports and groups which are the positions that can take the switch.
```
Comment (eckter) [inline api/osrd_infra/schemas/infra.py]:
```suggestion
    Switch types are used to define what kind of switch is used in the infrastructure.
    A switch type is defined by a list of ports and groups which are the possible configurations of the switch.
```

File: api/osrd_infra/schemas/infra.py [hunk 14]
```diff
@@ -258,8 +260,9 @@ class TrackSectionLink(BaseObjectTrait):
 
 
 class SpeedSection(BaseObjectTrait):
-    """This class is used to define speed sections.
-    Speed sections is recognized by its identifier and are in meters per second."""
+    """
+    Speed sections is recognized by its identifier and are in meters per second.
```
Comment (eckter) [inline api/osrd_infra/schemas/infra.py]:
```suggestion
    Speed sections are recognized by their identifiers and are in meters per second.
```

File: api/osrd_infra/schemas/infra.py [hunk 15]
```diff
@@ -271,14 +274,12 @@ class SpeedSection(BaseObjectTrait):
 
 
 class Catenary(BaseObjectTrait):
-    """This class is used to define catenary.
-    A catenary correspond at a set of cables designed
-    to power electric trains by capturing the current through
-    the use of a pantograph.
-    Catenary is identified by its identifier.
+    """
+    A catenary correspond at a set of cables designed to power electric trains by capturing the current through the use
```
Comment (eckter) [inline api/osrd_infra/schemas/infra.py]:
```suggestion
    A catenary corresponds to a set of cables designed to power electric trains by capturing the current through the use
```

File: api/osrd_infra/schemas/infra.py [hunk 16]
```diff
@@ -271,14 +274,12 @@ class SpeedSection(BaseObjectTrait):
...
+    of a pantograph. Catenary is identified by its identifier.
     """
 
-    voltage: float = Field(description="Type of power supply (in Volts) used for electrification", gt=0)
+    voltage: str = Field(description="Type of power supply (in Volts) used for electrification")
```
Comment (eckter) [inline api/osrd_infra/schemas/infra.py:282]: What is the reason for this change? `float` over `int` was probably a mistake, but `str` doesn't really seem fitting

File: api/osrd_infra/schemas/rolling_stock.py [hunk 17]
```diff
@@ -29,9 +28,47 @@ class EffortCurve(BaseModel, extra=Extra.forbid):
...
+
+class EffortCurveConditions(BaseModel, extra=Extra.forbid):
+    comfort: Optional[Union[Literal["ac"], Literal["heating"]]]
+
+
+class ConditionalEffortCurve(BaseModel, extra=Extra.forbid):
+    """Effort curve subject to application conditions"""
+
+    cond: EffortCurveConditions = Field(default=EffortCurveConditions())
+    curve: EffortCurve = Field(description="Effort curve to apply if the conditions are met")
```
Comment (eckter) [inline api/osrd_infra/schemas/rolling_stock.py]: It seems weird that the condition itself is not optional and its only parameter is optional. Is there a reason not to use an optional condition with a non-optional parameter?

File: core/src/main/java/fr/sncf/osrd/infra/api/tracks/undirected/TrackEdge.java [hunk 18]
```diff
@@ -42,5 +42,5 @@ public sealed interface TrackEdge permits SwitchBranch, TrackSection {
     ImmutableRangeMap<Double, LoadingGaugeConstraint> getLoadingGaugeConstraints();
 
     /** Returns a set of voltages usable at any position */
-    RangeMap<Double, Set<Integer>> getVoltages();
+    RangeMap<Double, String> getVoltages();
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/infra/api/tracks/undirected/TrackEdge.java]: How does this work? Don't we need a set of values? If we don't, the comment needs to be updated

File: api/openapi.yaml [hunk 19]
```diff
@@ -1061,26 +1063,62 @@ components:
...
+          type: object
+          nullable: true
+          properties:
+            comfort:
+              $ref: "#/components/schemas/Comfort"
+        curve:
+          $ref: "#/components/schemas/EffortCurve"
+    Comfort:
+      type: string
+      enum: [ "ac", "heating", "standard"]
```
Comment (multun) [inline api/openapi.yaml]: lowercase does not seem consistent with other enums in railjson
that's pretty minor though, I can live with it

File: api/osrd_infra/schemas/infra.py (hunk 16 above)
Comment (flomonster) [reply api/osrd_infra/schemas/infra.py:282]: Following this page: https://en.wikipedia.org/wiki/List_of_railway_electrification_systems

You can have for the same voltage a different frequency. 
Using a string is useful to model these cases and to match the rolling stock mode that can be `"thermal"`, `"1500"` etc...

File: core/src/main/java/fr/sncf/osrd/infra/api/tracks/undirected/TrackEdge.java (hunk 18 above)
Comment (flomonster) [reply core/src/main/java/fr/sncf/osrd/infra/api/tracks/undirected/TrackEdge.java]: We don't need a set since you cannot have multiple voltages at the same infra location. This would require having several physical catenaries at the same location.

File: api/osrd_infra/schemas/rolling_stock.py (hunk 17 above)
Comment (flomonster) [reply api/osrd_infra/schemas/rolling_stock.py]: In the future we will have multiple condition in `EffortCurveConditions` all of them being optional. However, having a `null` value for `cond` doesn't make sense since it's the same as using the `default_curve`.

File: core/src/main/java/fr/sncf/osrd/infra/implementation/tracks/undirected/TrackSectionImpl.java [hunk 20]
```diff
@@ -104,7 +104,7 @@ public ImmutableRangeMap<Double, LoadingGaugeConstraint> getLoadingGaugeConstrai
         return loadingGaugeConstraints;
     }
 
-    public RangeMap<Double, Set<Integer>> getVoltages() {
+    public RangeMap<Double, String> getVoltages() {
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/infra/implementation/tracks/undirected/TrackSectionImpl.java:107]: I guess it implies catenary sections cannot overlap; it makes total sense that it shouldn't happen (even though we should probably check with railway folks), but then what do we do when it does, raise a warning? (I think that's the way we should go)

Maybe it's time to rethink how to implement electrification, and make it a direct attribute of track sections. This way, we can easily detect / resolve overlaps during edition.

**Review by multun — DISMISSED** [review]

**Comment by multun** [comment]
Congrats to everyone involved! The backend work looks pretty good, I'm glad it didn't take too much work to integrate it all into the simulator.
Please rework the history before merging, I'd rather not have this many commits land for such an intertwined change.

File: core/src/main/java/fr/sncf/osrd/infra/implementation/tracks/undirected/TrackSectionImpl.java (hunk 20 above)
Comment (flomonster) [reply core/src/main/java/fr/sncf/osrd/infra/implementation/tracks/undirected/TrackSectionImpl.java:107]: Indeed having electrification as a direct track section attribute seems a better idea. However, changing that will break track section edtion so I prefer doing that change later

Comment (multun) [reply core/src/main/java/fr/sncf/osrd/infra/implementation/tracks/undirected/TrackSectionImpl.java:107]: oh yeah sure I didn't meant to advocate for doing it now

File: api/openapi.yaml (hunk 19 above)
Comment (flomonster) [reply api/openapi.yaml]: Done

**Review by multun — APPROVED** [review]

**Review by eckter — APPROVED** [review]

**Review by alexandredamiron — APPROVED** [review]
Perf improvement on next iteration are a priority
